pkg/build/controller/build/build_controller.go

```diff
@@ -388,25 +397,52 @@ func (bc *BuildController) resolveOutputDockerImageReference(build *buildapi.Bui
 			var ok bool
 			streamName, tag, ok = imageapi.SplitImageStreamTag(streamName)
 			if !ok {
-				return "", fmt.Errorf("the referenced image stream tag is invalid: %s", outputTo.Name)
+				return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced image stream tag is invalid: %s", outputTo.Name)
 			}
 			tag = ":" + tag
 		}
 		stream, err := bc.imageStreamStore.ImageStreams(namespace).Get(streamName)
 		if err != nil {
 			if errors.IsNotFound(err) {
-				return "", fmt.Errorf("the referenced output image stream %s/%s does not exist", namespace, streamName)
+				return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced output image stream %s/%s does not exist", namespace, streamName)
 			}
-			return "", fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
+			return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
 		}
 		if len(stream.Status.DockerImageRepository) == 0 {
 			e := fmt.Errorf("the image stream %s/%s cannot be used as the output for build %s/%s because the integrated Docker registry is not configured and no external registry was defined", namespace, outputTo.Name, build.Namespace, build.Name)
 			bc.recorder.Eventf(build, kapi.EventTypeWarning, "invalidOutput", "Error starting build: %v", e)
-			return "", e
+			return "", build.Spec.Output.PushSecret, e
 		}
 		ref = fmt.Sprintf("%s%s", stream.Status.DockerImageRepository, tag)
 	}
-	return ref, nil
+
+	pushSecret = build.Spec.Output.PushSecret
+	if pushSecret == nil {
+		serviceAccount := build.Spec.ServiceAccount
+		if len(serviceAccount) == 0 {
+			serviceAccount = bootstrappolicy.BuilderServiceAccountName
+		}
+		sa, err := bc.kubeClient.Core().ServiceAccounts(build.Namespace).Get(serviceAccount, metav1.GetOptions{})
+		if err != nil {
+			return ref, build.Spec.Output.PushSecret, fmt.Errorf("Error getting push/pull secrets for service account %s/%s: %v", build.Namespace, serviceAccount, err)
+		}
+
+		var builderSecrets []kapi.Secret
+		for _, saSecret := range sa.Secrets {
+			secret, err := bc.kubeClient.Core().Secrets(build.Namespace).Get(saSecret.Name, metav1.GetOptions{})
+			if err != nil {
+				continue
+			}
+			builderSecrets = append(builderSecrets, *secret)
+		}
+
+		pushSecret = buildutil.FindDockerSecret(builderSecrets, ref)
```

Comment:
OK, I got from the code that if the push secret is null we default to the build spec SA secret, but what is this translation from the builderSecret to the (docker?) push secret.  Does a quick comment of some sort make sense?